### DISCUSSION TITLE
Powershell Core i AD module

### DISCUSSION THREAD

**Discussion by RogalaPiotr** [AzureCommunityPL/group#31] [Pomoc] [answered]
Pytanie do naszego szanownego grona grupy Azure: Czy ktos probowal uruchomic modul do AAD (Azure AD) w Powershell Core? Znalazlem jeden opis dla "testowego" modulu o nazwie: "AzureAD.Standard.Preview" aczkolwiek sa z nim problemy na Mac'u mianowicie jakby nie chcial zaimportowac comandlet'ów, a modul jest zaimportowany.
Poszukuje tak zwanego pięknie w języku IT "workaround".
Z góry dzięki za podpowiedzi.

**Comment by eklime** [comment]
Hehe, moduł do AAD jest w wersji Desk, a nie Core, więc...no niestety 👎 
![image](https://user-images.githubusercontent.com/12485979/91171275-806b5c00-e6da-11ea-97b1-13fd1810f8a0.png)

**Reply by RogalaPiotr** [reply]
Tutaj jest opis gdzie niby dziala z AzureAD.Standard.Preview: https://endjin.com/blog/2019/05/how-to-use-the-azuread-module-in-powershell-core

Te wszystkie przetestowalem i wszystkie pojawiaja sie jako Desk, dlatego podpytuje bo moze ktos sie zmierzyl z problemem i sie udalo.

@eklime - alert, dokumentacja mowi ze nie powinno miec sie zainstalowanych obu modulow AzureAD i AzureADPreview.
https://docs.microsoft.com/en-us/powershell/azure/active-directory/install-adv2?view=azureadps-2.0

btw. dzieki za sprawdzenie :)

**Comment by RogalaPiotr** [comment]
Szybki update
znalazlem to: https://www.michev.info/Blog/Post/2339/azure-ad-powershell-module-with-support-for-powershell-core

Faktycznie jak wezmiesz ten modul z shella i dodasz sobie to dziala ale po zainstalowaniu .NETCore na macu.
https://dotnet.microsoft.com/download

Natomiast nadal mam problem z biblioteka logowania OAuth poniewaz nie mam wszystkich: ExportedCommands jak w shellu. Co powoduje ze modul jest widoczny z komendami jak na screenie ale brakuje komendy do zalogowania czyli: Connect-AzureAD.

![image](https://user-images.githubusercontent.com/31566956/91190778-3f804100-e6f4-11ea-8883-b3841a56df94.png)

**Reply by wi5nia** [reply]
Czyli generalnie jest, ale nie dziala wiec i tak bezuzyteczne ;)

**Reply by ptrstpp950** [reply]
Niestety z doświadczenia działa średnio na jeża. Jak już skonfigurujesz `AzureAD.Standard.Preview` to się okazuje, że jest tak stare, że nie kompatybilne. Jak weźmiesz `AzureADPreview` to podczas ładowania (`Import-Module`) dostaniesz błąd: `Could not load file or assembly 'System.Windows.Forms`. U mnie skończyło się to szybkim przepisaniem kawałka na `az ad`, które jest działa na różnych platformach.

**Comment by TomaszOledzki** [comment, marked as answer]
@RogalaPiotr AzureAd to moduł pisany w .Net Framework dla PowerShell Desktop pisanego w .Net Framework. PowerShell Core jest napisane w .Net Core. .Net Framework i .Net Core nie są zamienne. Nie odpalisz w PowerShell Core, modułu pisanego w .Net Framework. Czyli w PowerShell Core nie odpalisz moduł AzureAd. Musiałbyś skompilować sobie PowerShell Desktop w Mono (takie .Net Framework co działa na Maku), aby odpalić w nim moduł AzureAD na Maku :).

Potrzebujesz modułu pisanego dla PowerShell Core.

Do AAD możesz strzelać cmdletami z Az.Resources:
https://docs.microsoft.com/en-us/powershell/module/az.resources/?view=azps-4.6.0#active-directory. Najlepiej zainstaluj cały moduł Az, który już jest pisany pod PowerShell Core (Install-Module Az). Masz tam Connect-AzAccount.

Albo strzelaj bezpośrednio do Graph Api: https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-graph-api

**Reply by RogalaPiotr** [reply]
dzieki za hint z Graph Api 👍 

W zasadzie mam kilka tematów wokół tego zagadnienia:
1. Shell w Portalu Azure dziala na PowerShell Core na linuxie gdzie jest modul AzureAD.Standard.Preview w wersji która jest nie dostepna w PS Gallery testowej. Dlatego szukam opcji uruchomienia tego na tych samych zasadach na Macu(dla wlasnej wygody)
2. AzureAD jak napisales nie jest dla Core dlatego go nie testuje nawet heh, ale jest ADPreview ktory w zasadzie jest rekomendowany w dokumentacji i chyba jest dla Core tyle ze sa z nim nadal problemy.
3. Obecnie modul AzureAD jest mi potrzebny do PIMa poniewaz on wymaga zalogowania sie do AD.
4. Temat chwilowo porzuce ze wzgledu na to ze potrzebuje miec rozwiazanie ktore bedzie dzialac rowniez w AAD Governmentowym a juz wiem ze beda z tym problemy. Dlatego pozostaje zabawa po API.
5. Ogarniecie modulu AD dla Core pozwoliloby mi rowniez na prosta implemenetacje do ADO nie zmieniajac skryptu i rozbijac go na rozne wesje PSa itd.

Mysle ze w najblizszym czasie zostanie wydana nowa wersja modulu wiec problem sie rozwiaze. Dzieki za pomoc.

**Reply by wi5nia** [reply]
Embrace yourself... Ignite is coming!

**Reply by TomaszOledzki** [reply]
@RogalaPiotr próbowałeś uwierzytelnić się przez Connect-AzAccount i użyć cmdlet z AzureAdPreview? Może podbierze sobie JWT z bieżącej sesji. W sumie to ciekawi mnie czy to zadziała, ale chwilowo jestem odcięty od Maka i sam nie sprawdzę :(.

Update:
Chociaż w sumie może nie działać. Connect-AzAccount uwierzytelnia się w ARM, a ADPreview pewnie w Graph Api.

**Reply by RogalaPiotr** [reply]
@TomaszOledzki napisales zebym przetestowal zalogowanie Connect-AzAccount - kiedy chcialem to zrobic wyskoczyl mi error ze jest jakis problem z zaladowniem "Connect-AzAccount: Could not load type 'System.Security.Cryptography.SHA256Cng' from assembly 'System.Core,". Wiec zrestartowalem konsole zeby sie upewnic czy nie jest to blad w wyniku importowania tych modulow. Po restarcie normalnie sie zalogowalem uzywajac Login-AzAccount oraz Connect-AzAccount - i dziala poprawnie. Zaimportowalem sobie ponownie AzureAD.Standard.Preview - sciagniety z Shella w wersji 0.0.0.10 - i sie mocno zdzwilem poniewaz pojawila sie komenda Connect-AzureAD - i w dodatku dziala.

![image](https://user-images.githubusercontent.com/31566956/91283516-fa5b1e00-e78a-11ea-8ba9-398b554fdefd.png)


Jesli ktos jest zainteresowany zainstalowaniem sobie AzureAD dla PS Core to nalezy wykonac kolejne kroki:

Pobrac modul z shella: https://www.michev.info/Blog/Post/2339/azure-ad-powershell-module-with-support-for-powershell-core i wrzucic sobie na dysk gdzie sa moduly do PS
Zainstalowac .NETCore w wersji rekomendowanej: 3.1
Zrestartowac kompa 3 razy i zaimportowac modul - powinno dzialac.

**Comment by KamilWiecek** [comment]
Ja uruchamiam to w pipelinie z agentem Windowsowym. Niestety jedyny sposób na wykorzystanie Service Principal wiąże się koniecznością ściągnięcia i zaimportowania certyfikatu np. z KV. (https://docs.microsoft.com/en-us/powershell/azure/active-directory/signing-in-service-principal?view=azureadps-2.0). Wymagane są oczywiście odpowiednie API Permissions.

```powershell
[CmdletBinding()]
param (
    [Parameter()]
    [string]
    $VaultName,

    [Parameter()]
    [string]
    $CertName
)

$context = Get-AzContext

$cert = Get-AzKeyVaultCertificate  -VaultName $VaultName -Name $CertName
$secret = Get-AzKeyVaultSecret -VaultName $VaultName -Name $cert.Name

$secretByte = [Convert]::FromBase64String($secret.SecretValueText)
$x509Cert = new-object System.Security.Cryptography.X509Certificates.X509Certificate2
$x509Cert.Import($secretByte, "", "Exportable,PersistKeySet")

$password = (New-Guid).Guid
$secPassword = $password | ConvertTo-SecureString -AsPlainText -Force

$type = [System.Security.Cryptography.X509Certificates.X509ContentType]::Pfx
$pfxFileByte = $x509Cert.Export($type, $password)
$exportPath = Join-Path -Path (Get-Location) -ChildPath 'export.pfx'
[System.IO.File]::WriteAllBytes($exportPath, $pfxFileByte)
Import-PfxCertificate -Password $secPassword -FilePath $exportPath -CertStoreLocation Cert:\LocalMachine\My

Write-Output "Imported cert"

Remove-Item -Path $exportPath -Force

Connect-AzureAD -CertificateThumbprint $cert.Thumbprint -ApplicationId $context.Account.Id -TenantId $context.Tenant.Id

Write-Output "Connected to Azure AD"

# Twoja logika ...

```
